### PR TITLE
feat/support_multicast_decoding

### PR DESCRIPTION
**Pull request by Haiss2**
Support multicast decoding for deribit-api

### CODE REVIEW

File: pkg/multicast/common.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package multicast
+
+func newInstrumentNotificationChannel() string {
+	return "instrument"
```
Comment (hiepnv90) [inline pkg/multicast/common.go]: Should be `instrument.state.<kind>.<currency>` (likes websocket subscription)

File: pkg/multicast/common.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+func newOrderBookNotificationChannel(instrument string) string {
+	return "book." + instrument
+}
+
+func newTradesNotificationChannel(instrument string) string {
+	return "trades." + instrument
```
Comment (hiepnv90) [inline pkg/multicast/common.go]: `trades.<kind>.<currency>`

File: .test.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Question?
```
Comment (hiepnv90) [inline .test.txt]: Delete file

File: pkg/multicast/multicast.go [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+// Event represents a Deribit multicast events.
+type Event struct {
+	Instrument string
+	Type       EventType
+	Data       interface{}
+}
+
+// Client represents a client for Deribit multicast.
+type Client struct {
+	m           *sbe.SbeGoMarshaller
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't use this object for multiple go-routines.

File: pkg/multicast/multicast.go [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+	log         *zap.SugaredLogger
+	mu          sync.RWMutex
+	inf         *net.Interface
+	addrs       []string
+	addrConnMap map[string]*net.UDPConn
+	wsClient    *websocket.Client
+
+	isConnected        bool
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: UDP is connectionless.

File: pkg/multicast/multicast.go [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+		wsClient:    wsClient,
+		addrConnMap: make(map[string]*net.UDPConn),
+
+		supportCurrencies:  currencies,
+		instrumentIDToName: make(map[uint32]string),
+		channelIDSeqNum:    make(map[uint16]uint32),
+		emitter:            emission.NewEmitter(),
+	}
+
+	err = client.buildInstrumentIDToNameMapping()
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't build this mapping here.

File: pkg/multicast/multicast.go [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+	if err != nil {
+		l.Errorw("failed to build instrumentId name mapping", "err", err)
+		return nil, err
+	}
+
+	err = client.Start()
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't start client here. Just return the client object.

File: pkg/multicast/multicast.go [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+	case 1000:
+		return c.decodeInstrumentEvent(r, header)
+	case 1001:
+		return c.decodeOrderBookEvent(r, header)
+	case 1002:
+		return c.decodeTradesEvent(r, header)
+	case 1003:
+		return c.decodeTickerEvent(r, header)
+	default:
+		return Event{}, nil
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should return error for unsupported `templateId`

File: pkg/multicast/multicast.go [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+func (c *Client) decodeOrderBookEvent(r io.Reader, header sbe.MessageHeader) (Event, error) {
+	var b sbe.Book
+	err := b.Decode(c.m, r, header.BlockLength, true)
+	if err != nil {
+		return Event{}, nil
+	}
+
+	c.mu.Lock()
+	instrumentName := c.instrumentIDToName[b.InstrumentId]
+	c.mu.Unlock()
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: This code snippet should be wrap into a function for other use.

File: pkg/multicast/multicast.go [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+	if err != nil {
+		c.log.Errorw("failed to decode events", "err", err)
+		return err
+	}
+
+	for _, event := range events {
+		switch event.Type {
+		case EventTypeInstrument:
+			// update instrumentIDToName mapping
+			var ins models.Instrument
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]:
```go
ins := event.Data.(models.Instrument)
```

File: pkg/multicast/multicast.go [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+const (
+	EventTypeInstrument EventType = iota
+	EventTypeOrderBook
+	EventTypeTrades
+	EventTypeTicker
+)
+
+// Event represents a Deribit multicast events.
+type Event struct {
+	Instrument string
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need this field.

File: pkg/multicast/multicast.go [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+	}
+	return Event{
+		Instrument: instrumentName,
+		Type:       EventTypeTicker,
+		Data:       ticker,
+	}, nil
+}
+
+// IsConnected checks whether the connection is established or not.
+func (c *Client) IsConnected() bool {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need this check.

File: pkg/multicast/multicast.go [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.isConnected
+}
+
+// Start starts listening events on interface `ifname` and `addrs`.
+func (c *Client) Start() error {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should pass context to this method

File: pkg/multicast/multicast.go [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+		err = udpConn.SetReadBuffer(defaultReadBufferSize)
+		if err != nil {
+			c.log.Errorw("fail to set read buffer for UDP", "error", err)
+			return err
+		}
+
+		c.addrConnMap[addr] = udpConn
+	}
+
+	return c.ListenToEvents()
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Pass context to this function

File: pkg/multicast/multicast.go [hunk 15]
```diff
@@ -0,0 +1,510 @@
...
+
+	lastSeq, ok := c.getSeqNum(channelID)
+	if !ok {
+		c.setSeqNum(channelID, seq)
+		return nil
+	}
+
+	// jumped pkg
+	if (seq > lastSeq && seq-lastSeq > 1) || (lastSeq == math.MaxUint32 && seq > 0) {
+		return fmt.Errorf("missing package, current seq: %d, last seq: %d", seq, lastSeq)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should define an error for this type of error, e.g: ErrLostPackage

File: pkg/multicast/multicast.go [hunk 16]
```diff
@@ -0,0 +1,510 @@
...
+	}
+
+	// connection reset
+	if lastSeq < math.MaxInt32 && seq == 0 {
+		return fmt.Errorf("connection reset, last seq: %d", lastSeq)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should define an error this this type of error: `ErrConnectionReset`

File: pkg/multicast/multicast.go [hunk 17]
```diff
@@ -0,0 +1,510 @@
...
+			}
+
+			err = json.Unmarshal(dataBytes, &ins)
+			if err != nil {
+				c.log.Errorw("failed to decode event data", "err", err)
+				return err
+			}
+			c.mu.Lock()
+			c.instrumentIDToName[ins.InstrumentID] = ins.InstrumentName
+			c.mu.Unlock()
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Write a function for set this map value.

File: pkg/multicast/multicast.go [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+			c.Emit(newInstrumentNotificationChannel(), event)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should emit `event.Data`.

File: pkg/multicast/multicast.go [hunk 19]
```diff
@@ -0,0 +1,510 @@
...
+
+		case EventTypeOrderBook:
+			c.Emit(newOrderBookNotificationChannel(event.Instrument), event)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should emit `event.Data`

File: pkg/multicast/multicast.go [hunk 20]
```diff
@@ -0,0 +1,510 @@
...
+
+		case EventTypeTrades:
+			c.Emit(newTradesNotificationChannel(event.Instrument), event)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should emit `event.Data`

File: pkg/multicast/multicast.go [hunk 21]
```diff
@@ -0,0 +1,510 @@
...
+
+		case EventTypeTicker:
+			c.Emit(newTickerNotificationChannel(event.Instrument), event)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should emit `event.Data`

File: pkg/multicast/multicast.go [hunk 22]
```diff
@@ -0,0 +1,510 @@
...
+
+		default:
+			return fmt.Errorf("invalid event type: %v", event.Type)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Ignore the unknown event type

File: pkg/multicast/multicast.go [hunk 23]
```diff
@@ -0,0 +1,510 @@
...
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) getConnection(addr string) *net.UDPConn {
+	return c.addrConnMap[addr]
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:465]: Use lock here

File: pkg/multicast/multicast.go [hunk 24]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+// ListenToEventsForAddress listens to one udp address on given network interface.
+// If it receives an InstrumentChange event, it has to update the instrumentIDToName accordingly.
+func (c *Client) ListEventsForAddress(addr string) error {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: This function should setup an new UDP connection and use it, also allocate a Sbe marshaller to use for this connection only.

File: pkg/multicast/multicast.go [hunk 25]
```diff
@@ -0,0 +1,510 @@
...
+	// listen to event and then handle
+	for {
+		data := make([]byte, defaultReadBufferSize)
+		n, err := conn.Read(data)
+		if err != nil {
+			c.log.Errorw("fail to read UDP package", "error", err)
+			return err
+		} else {
+			toBuffer := bytes.NewBuffer(data[:n])
+			err := c.Handle(toBuffer)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Handle for package lost and connection reset error. Currently, we treat them the same, need to restart the connection.

File: pkg/multicast/multicast.go [hunk 26]
```diff
@@ -0,0 +1,510 @@
...
+// ListenToEventsForAddress listens to one udp address on given network interface.
+// If it receives an InstrumentChange event, it has to update the instrumentIDToName accordingly.
+func (c *Client) ListEventsForAddress(addr string) error {
+	conn := c.getConnection(addr)
+	defer conn.Close()
+
+	// listen to event and then handle
+	for {
+		data := make([]byte, defaultReadBufferSize)
+		n, err := conn.Read(data)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should read the data from the connection in a separate go routine (use for loop to read util the error occur) and put the data to a channel for later processing.

File: pkg/multicast/multicast.go [hunk 27]
```diff
@@ -0,0 +1,510 @@
...
+		if err != nil {
+			c.log.Errorw("fail to read UDP package", "error", err)
+			return err
+		} else {
+			toBuffer := bytes.NewBuffer(data[:n])
+			err := c.Handle(toBuffer)
+			c.log.Errorw("fail to handle UDP package", "error", err)
+			return err
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't return error here.

File: pkg/multicast/multicast.go [hunk 28]
```diff
@@ -0,0 +1,545 @@
...
+
+func (c *Client) setConnection(addr string, conn *net.UDPConn) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.addrConnMap[addr] = conn
+}
+
+// ListenToEventsForAddress listens to one udp address on given network interface.
+func (c *Client) ListEventsForAddress(ctx context.Context, addr string) error {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: `ListenToEventsForAddress`

File: pkg/multicast/multicast.go [hunk 29]
```diff
@@ -0,0 +1,545 @@
...
+	}
+
+	lastSeq, ok := c.getSeqNum(channelID)
+	if !ok {
+		c.setSeqNum(channelID, seq)
+		return nil
+	}
+
+	// jumped pkg
+	if (seq > lastSeq && seq-lastSeq > 1) || (lastSeq == math.MaxUint32 && seq > 0) {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: What about `seq < lastSeq && (math.MaxUint32 - lastSeq + seq > 1)`

Comment (hiepnv90) [reply pkg/multicast/multicast.go]: Check for valid sequence number will be easier.

File: pkg/multicast/multicast.go [hunk 30]
```diff
@@ -0,0 +1,545 @@
...
+		case EventTypeInstrument:
+			// update instrumentsMap
+			ins, ok := event.Data.(models.Instrument)
+			if !ok {
+				return fmt.Errorf("invalid event type")
+			}
+			c.setInstrument(ins.InstrumentID, ins)
+
+			// emit event
+			for _, currency := range c.supportCurrencies {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need this for loop. Can get currency from instrument: `currency = ins.BaseCurrency`.

File: pkg/multicast/multicast.go [hunk 31]
```diff
@@ -0,0 +1,545 @@
...
+		case EventTypeOrderBook:
+			books := event.Data.(models.OrderBookNotification)
+			c.Emit(newOrderBookNotificationChannel(books.InstrumentName), event.Data)
+
+		case EventTypeTrades:
+			trades := event.Data.(models.TradesNotification)
+			if len(trades) > 0 {
+				tradeIns := trades[0].InstrumentName
+				tradeKind := trades[0].InstrumentKind
+				for _, currency := range c.supportCurrencies {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need this for loop. Can get currency from instrument: currency = ins.BaseCurrency.

File: pkg/multicast/multicast.go [hunk 32]
```diff
@@ -0,0 +1,545 @@
...
+	if err != nil {
+		c.log.Errorw("failed to listen to multicast UDP", "error", err)
+		return err
+	}
+
+	err = udpConn.SetReadBuffer(defaultReadBufferSize)
+	if err != nil {
+		c.log.Errorw("fail to set read buffer for UDP", "error", err)
+		return err
+	}
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:492]: Write a function `listenMulticastUDP` that do all the above steps.

File: pkg/multicast/multicast.go [hunk 33]
```diff
@@ -0,0 +1,545 @@
...
+
+	// handle data from dataCh
+	go func() {
+		m := sbe.NewSbeGoMarshaller()
+		for {
+			select {
+			case <-ctx.Done():
+				udpConn.Close()
+			case data := <-dataCh:
+				toBuffer := bytes.NewBuffer(data)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Variable name should be a noun

File: pkg/multicast/multicast.go [hunk 34]
```diff
@@ -0,0 +1,545 @@
...
+				err := c.Handle(m, toBuffer)
+				if errors.Is(err, ErrConnectionReset) || errors.Is(err, ErrLostPackage) {
+					c.log.Infow("connection reset or lost package err, restarting connection...", "error", err)
+					err := c.restartConnection(ctx, addr)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:519]: Restart connection in a new go-routine and stop this go-routine.

File: pkg/multicast/multicast.go (hunk 31 above)
Comment (Haiss2) [reply pkg/multicast/multicast.go]: models.Trade doesn't contain models.Instrument, should I add the `BaseCurrency` field into the models.Trade?

File: pkg/multicast/multicast.go [hunk 35]
```diff
@@ -0,0 +1,548 @@
...
+	conn := c.getConnection(addr)
+	err := conn.Close()
+	if err != nil {
+		c.log.Errorw("failed to close connection", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (c *Client) restartConnection(ctx context.Context, addr string) error {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:315]: restartConnection should re-build instrumentMap.

**Review by hiepnv90 — APPROVED** [review]
LGTM! :+1:

**Pull request merged**